<a href="https://colab.research.google.com/github/Roseyshi/Liver-Cirrhosis-Classification/blob/main/Liver_Cirrhosis_Preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **LIVER_CIRRHOSIS CLASSIFICATION**

### **PROBLEM STATEMENT**

### **OBJECTIVES**

### **DATA UNDERSTANDING**

In [1]:
#Import libraries and dependencies
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')


In [2]:
#Load and read the dataset
data = pd.read_csv('/content/liver_cirrhosis.csv')
data.head()

N_Days Status     Drug    Age Sex Ascites Hepatomegaly Spiders Edema  \
0    2221      C  Placebo  18499   F       N            Y       N     N   
1    1230      C  Placebo  19724   M       Y            N       Y     N   
2    4184      C  Placebo  11839   F       N            N       N     N   
3    2090      D  Placebo  16467   F       N            N       N     N   
4    2105      D  Placebo  21699   F       N            Y       N     N   

   Bilirubin  Cholesterol  Albumin  Copper  Alk_Phos    SGOT  Tryglicerides  \
0        0.5        149.0     4.04   227.0     598.0   52.70           57.0   
1        0.5        219.0     3.93    22.0     663.0   45.00           75.0   
2        0.5        320.0     3.54    51.0    1243.0  122.45           80.0   
3        0.7        255.0     3.74    23.0    1024.0   77.50           58.0   
4        1.9        486.0     3.54    74.0    1052.0  108.50          109.0   

   Platelets  Prothrombin  Stage  
0      256.0          9.9      1  
1      220.0         10.8      2  
2      225.0         10.0      2  
3      151.0         10.2      2  
4      151.0         11.5      1

In [8]:
data.shape

(25000, 19)

The data has 25000 rows and 19 columns.

In [12]:
#Understand the columns
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 19 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   N_Days         25000 non-null  int64  
 1   Status         25000 non-null  object 
 2   Drug           25000 non-null  object 
 3   Age            25000 non-null  int64  
 4   Sex            25000 non-null  object 
 5   Ascites        25000 non-null  object 
 6   Hepatomegaly   25000 non-null  object 
 7   Spiders        25000 non-null  object 
 8   Edema          25000 non-null  object 
 9   Bilirubin      25000 non-null  float64
 10  Cholesterol    25000 non-null  float64
 11  Albumin        25000 non-null  float64
 12  Copper         25000 non-null  float64
 13  Alk_Phos       25000 non-null  float64
 14  SGOT           25000 non-null  float64
 15  Tryglicerides  25000 non-null  float64
 16  Platelets      25000 non-null  float64
 17  Prothrombin    25000 non-null  float64
 18  Stage 

**Statistics of the dataset**

The describe method shows a summary of statistical inferences of numeric columns. The information includes:


*   Count of entries in the columns.
*   Mean / average.
*   Standard Deviation.
*   Minimum and Maximum values.
*   25th, 50th(median) and 75th percentiles.







In [14]:
#Use the describe method
data.describe()

N_Days           Age     Bilirubin   Cholesterol       Albumin  \
count  25000.000000  25000.000000  25000.000000  25000.000000  25000.000000   
mean    1887.117040  18495.877080      3.402644    372.331471      3.486578   
std     1091.690918   3737.596616      4.707491    193.668452      0.380488   
min       41.000000   9598.000000      0.300000    120.000000      1.960000   
25%     1080.000000  15694.000000      0.800000    275.000000      3.290000   
50%     1680.000000  18499.000000      1.300000    369.510563      3.510000   
75%     2576.000000  20955.000000      3.400000    369.510563      3.750000   
max     4795.000000  28650.000000     28.000000   1775.000000      4.640000   

             Copper      Alk_Phos          SGOT  Tryglicerides     Platelets  \
count  25000.000000  25000.000000  25000.000000   25000.000000  25000.000000   
mean     100.184663   1995.675597    123.166345     123.822548    256.007337   
std       73.184840   1798.885660     47.747616      52.786350     98.679006   
min        4.000000    289.000000     26.350000      33.000000     62.000000   
25%       52.000000   1032.000000     92.000000      92.000000    189.000000   
50%       97.648387   1828.000000    122.556346     124.702128    251.000000   
75%      107.000000   1982.655769    134.850000     127.000000    311.000000   
max      588.000000  13862.400000    457.250000     598.000000    721.000000   

        Prothrombin        Stage  
count  25000.000000  25000.00000  
mean      10.734549      2.00116  
std        0.904436      0.81387  
min        9.000000      1.00000  
25%       10.000000      1.00000  
50%       10.600000      2.00000  
75%       11.100000      3.00000  
max       18.000000      3.00000

#### **Data Overview**

This section focuses on the attributes of the data. This encompasses:


*   Uniformity - Appropriate data types.
*   Completeness - Null values.
*   Consistency - Duplicated values.

This data properties are important in ensuring that we have data that is up to standard and reliable to obtain keny insights and findings, then use those findings train a model.





**Data Uniformity**

In [4]:
#Check uniformity in data types
data.dtypes

N_Days             int64
Status            object
Drug              object
Age                int64
Sex               object
Ascites           object
Hepatomegaly      object
Spiders           object
Edema             object
Bilirubin        float64
Cholesterol      float64
Albumin          float64
Copper           float64
Alk_Phos         float64
SGOT             float64
Tryglicerides    float64
Platelets        float64
Prothrombin      float64
Stage              int64
dtype: object

Data types are accordingly assigend to the columns. Columns that are in integer form are accordingly assgined so and also the category data columns.

**Data Completeness**

In [5]:
#Check for null values
data.isnull().sum()

N_Days           0
Status           0
Drug             0
Age              0
Sex              0
Ascites          0
Hepatomegaly     0
Spiders          0
Edema            0
Bilirubin        0
Cholesterol      0
Albumin          0
Copper           0
Alk_Phos         0
SGOT             0
Tryglicerides    0
Platelets        0
Prothrombin      0
Stage            0
dtype: int64

We do not have any null values in this dataset. No action will be taken in this regard.

**Data Consistency**

In [7]:
#Check for duplicates
data.duplicated().sum()

15361

There seems to be a be a very large number of duplicated rows. 15361 out of the 25000 rows. This would largely affect the prfromance of the model as it would be biased towards the frequently appearing stage category.

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(data.drop('Stage', axis=1), data['Stage'], test_size=0.2, random_state=42)

df_test = pd.concat([X_test, y_test], axis=1)
df_test.to_csv('test_data.csv', index=False)

# Optionally, save the training data if needed
df_train = pd.concat([X_train, y_train], axis=1)
df_train.to_csv('train_data.csv', index=False)

In [ ]:
train_data = pd.read_csv('/content/train_data.csv')
test_data = pd.read_csv('/content/test_data.csv')

In [ ]:
y_train.value_counts()

Stage
2    6744
3    6648
1    6608
Name: count, dtype: int64

In [ ]:
from sklearn.preprocessing import OneHotEncoder

encoder = OneHotEncoder(drop='first', sparse=False)

X_train_encoded = encoder.fit_transform(train_data[['Status', 'Drug', 'Sex', 'Ascites', 'Hepatomegaly', 'Spiders', 'Edema']])
X_test_encoded = encoder.transform(test_data[['Status', 'Drug', 'Sex', 'Ascites', 'Hepatomegaly', 'Spiders', 'Edema']])

X_train_encoded_df = pd.DataFrame(X_train_encoded, columns=encoder.get_feature_names_out())
X_test_encoded_df = pd.DataFrame(X_test_encoded, columns=encoder.get_feature_names_out())

train_data.reset_index(drop=True, inplace=True)
test_data.reset_index(drop=True, inplace=True)

X_train_encoded = pd.concat([train_data, X_train_encoded_df], axis=1)
X_test_encoded = pd.concat([test_data, X_test_encoded_df], axis=1)

X_train_encoded = X_train_encoded.drop(columns = ['Status','Drug','Sex','Ascites','Hepatomegaly','Spiders','Edema'], axis = 1)
X_test_encoded = X_test_encoded.drop(columns = ['Status','Drug','Sex','Ascites','Hepatomegaly','Spiders','Edema'], axis = 1)

X_train_encoded.head()


N_Days    Age  Bilirubin  Cholesterol  Albumin      Copper     Alk_Phos  \
0    2772  20955        0.6   217.000000     3.62   13.000000   414.000000   
1    2835  17809        0.6   296.000000     3.42   37.000000  1032.000000   
2    1236  18993        1.5   369.510563     3.42   97.648387  1982.655769   
3    1321  14610        0.8   328.000000     3.31   62.000000  1105.000000   
4    3574  24585        4.5   274.000000     4.09  154.000000  1533.000000   

         SGOT  Tryglicerides  Platelets  ...  Stage  Status_CL  Status_D  \
0   75.950000     119.000000      224.0  ...      2        0.0       0.0   
1   80.600000      83.000000      487.0  ...      1        0.0       0.0   
2  122.556346     124.702128      231.0  ...      2        0.0       0.0   
3  137.000000      95.000000      117.0  ...      3        0.0       0.0   
4  117.800000     128.000000      412.0  ...      2        0.0       1.0   

   Drug_Placebo  Sex_M  Ascites_Y  Hepatomegaly_Y  Spiders_Y  Edema_S  Edema_Y  
0           1.0    0.0        0.0             0.0        0.0      0.0      0.0  
1           1.0    0.0        0.0             0.0        0.0      0.0      0.0  
2           1.0    0.0        1.0             0.0        1.0      0.0      0.0  
3           1.0    0.0        1.0             0.0        1.0      0.0      0.0  
4           0.0    0.0        0.0             0.0        0.0      0.0      0.0  

[5 rows x 21 columns]

In [ ]:
X_test_encoded.head()

N_Days    Age  Bilirubin  Cholesterol  Albumin      Copper     Alk_Phos  \
0    3358  17167        2.1   262.000000     3.48   58.000000  2045.000000   
1    1443  14975        1.2   369.510563     2.80   97.648387  1982.655769   
2     694  18993        0.8   300.000000     2.94  231.000000  1794.000000   
3    2634  18972        0.9   346.000000     3.09   81.000000  1098.000000   
4    1492  14106        3.2   369.510563     3.56   77.000000  1790.000000   

         SGOT  Tryglicerides  Platelets  ...  Stage  Status_CL  Status_D  \
0   89.900000      84.000000      475.0  ...      1        0.0       1.0   
1  122.556346     124.702128      120.0  ...      1        0.0       0.0   
2  130.200000      99.000000      219.0  ...      3        0.0       1.0   
3  122.450000      90.000000      165.0  ...      3        0.0       0.0   
4  139.500000     124.702128      309.0  ...      3        0.0       1.0   

   Drug_Placebo  Sex_M  Ascites_Y  Hepatomegaly_Y  Spiders_Y  Edema_S  Edema_Y  
0           1.0    0.0        1.0             0.0        1.0      0.0      0.0  
1           1.0    0.0        1.0             0.0        1.0      0.0      0.0  
2           1.0    0.0        1.0             1.0        1.0      0.0      0.0  
3           0.0    0.0        0.0             1.0        1.0      1.0      0.0  
4           0.0    0.0        0.0             1.0        0.0      0.0      0.0  

[5 rows x 21 columns]

In [ ]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train_encoded)
X_test_scaled = scaler.transform(X_test_encoded)

In [ ]:
X_train_scaled = pd.DataFrame(X_train_scaled, columns = X_train_encoded.columns)
X_test_scaled = pd.DataFrame(X_test_scaled, columns = X_test_encoded.columns)

X_train_scaled.head()

N_Days       Age  Bilirubin  Cholesterol   Albumin    Copper  Alk_Phos  \
0  0.806410  0.653246  -0.596160    -0.805880  0.350795 -1.191836 -0.880537   
1  0.863929 -0.186407  -0.596160    -0.393963 -0.173327 -0.865377 -0.536696   
2 -0.595960  0.129597  -0.405786    -0.010668 -0.173327 -0.040410 -0.007772   
3 -0.518354 -1.040204  -0.553855    -0.227110 -0.461594 -0.525316 -0.496080   
4  1.538637  1.622075   0.228795    -0.508674  1.582482  0.726110 -0.257950   

       SGOT  Tryglicerides  Platelets  ...     Stage  Status_CL  Status_D  \
0 -0.988688      -0.093316  -0.323940  ... -0.002457  -0.296184 -0.782881   
1 -0.891349      -0.767540   2.347186  ... -1.230772  -0.296184 -0.782881   
2 -0.013072       0.013476  -0.252846  ... -0.002457  -0.296184 -0.782881   
3  0.289278      -0.542799  -1.410672  ...  1.225859  -0.296184 -0.782881   
4 -0.112637       0.075240   1.585458  ... -0.002457  -0.296184  1.277333   

   Drug_Placebo     Sex_M  Ascites_Y  Hepatomegaly_Y  Spiders_Y   Edema_S  \
0      0.757422 -0.361008  -0.697450        -0.80144  -0.918971 -0.417445   
1      0.757422 -0.361008  -0.697450        -0.80144  -0.918971 -0.417445   
2      0.757422 -0.361008   1.433795        -0.80144   1.088173 -0.417445   
3      0.757422 -0.361008   1.433795        -0.80144   1.088173 -0.417445   
4     -1.320268 -0.361008  -0.697450        -0.80144  -0.918971 -0.417445   

    Edema_Y  
0 -0.214535  
1 -0.214535  
2 -0.214535  
3 -0.214535  
4 -0.214535  

[5 rows x 21 columns]

In [ ]:
X_test_scaled.head()

N_Days       Age  Bilirubin  Cholesterol   Albumin    Copper  Alk_Phos  \
0  1.341429 -0.357753  -0.278870    -0.571243 -0.016091 -0.579726  0.026915   
1 -0.406968 -0.942788  -0.469244    -0.010668 -1.798106 -0.040410 -0.007772   
2 -1.090806  0.129597  -0.553855    -0.373106 -1.431220  1.773500 -0.112736   
3  0.680416  0.123993  -0.532702    -0.133255 -1.038129 -0.266869 -0.499975   
4 -0.362231 -1.174719  -0.046190    -0.010668  0.193558 -0.321279 -0.114961   

       SGOT  Tryglicerides  Platelets  ...     Stage  Status_CL  Status_D  \
0 -0.696671      -0.748812   2.225309  ... -1.230772  -0.296184  1.277333   
1 -0.013072       0.013476  -1.380203  ... -1.230772  -0.296184 -0.782881   
2  0.146933      -0.467885  -0.374722  ...  1.225859  -0.296184  1.277333   
3 -0.015298      -0.636441  -0.923166  ...  1.225859  -0.296184 -0.782881   
4  0.341611       0.013476   0.539351  ...  1.225859  -0.296184  1.277333   

   Drug_Placebo     Sex_M  Ascites_Y  Hepatomegaly_Y  Spiders_Y   Edema_S  \
0      0.757422 -0.361008   1.433795       -0.801440   1.088173 -0.417445   
1      0.757422 -0.361008   1.433795       -0.801440   1.088173 -0.417445   
2      0.757422 -0.361008   1.433795        1.247754   1.088173 -0.417445   
3     -1.320268 -0.361008  -0.697450        1.247754   1.088173  2.395526   
4     -1.320268 -0.361008  -0.697450        1.247754  -0.918971 -0.417445   

    Edema_Y  
0 -0.214535  
1 -0.214535  
2 -0.214535  
3 -0.214535  
4 -0.214535  

[5 rows x 21 columns]